packages/multichain/src/caip25Permission.test.ts

```diff
@@ -0,0 +1,696 @@
+import * as ScopeAssert from './scope/assert';
+import * as ScopeAuthorization from './scope/authorization';
+
+jest.mock('./scope/authorization', () => ({
+  validateAndFlattenScopes: jest.fn(),
+}));
+const MockScopeAuthorization = jest.mocked(ScopeAuthorization);
+
+jest.mock('./scope/assert', () => ({
+  assertScopesSupported: jest.fn(),
+}));
+const MockScopeAssert = jest.mocked(ScopeAssert);
+
+const { removeAccount } = Caip25CaveatMutatorFactories[Caip25CaveatType];
+
+describe('endowment:caip25', () => {
+  beforeEach(() => {
+    MockScopeAuthorization.validateAndFlattenScopes.mockReturnValue({
+      flattenedRequiredScopes: {},
+      flattenedOptionalScopes: {},
+    });
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it('builds the expected permission specification', () => {
+    const specification = caip25EndowmentBuilder.specificationBuilder({
+      methodHooks: {
+        findNetworkClientIdByChainId: jest.fn(),
+      },
+    });
+    expect(specification).toStrictEqual({
+      permissionType: PermissionType.Endowment,
+      targetName: Caip25EndowmentPermissionName,
+      endowmentGetter: expect.any(Function),
+      allowedCaveats: [Caip25CaveatType],
+      validator: expect.any(Function),
+    });
+
+    expect(specification.endowmentGetter()).toBeNull();
+  });
+
+  describe('caveat mutator removeScope', () => {
+    it('can remove a caveat', () => {
+      const ethereumGoerliCaveat = {
+        requiredScopes: {
+          'eip155:1': {
+            methods: ['eth_call'],
+            notifications: ['chainChanged'],
+          },
+        },
+        optionalScopes: {
+          'eip155:5': {
+            methods: ['eth_call'],
+            notifications: ['accountsChanged'],
+          },
+        },
+        sessionProperties: {},
+        isMultichainOrigin: true,
+      };
+      const result = removeScope('eip155:5', ethereumGoerliCaveat);
+      expect(result).toStrictEqual({
+        operation: CaveatMutatorOperation.UpdateValue,
+        value: {
+          requiredScopes: {
+            'eip155:1': {
+              methods: ['eth_call'],
+              notifications: ['chainChanged'],
+            },
+          },
+          optionalScopes: {},
+        },
+      });
+    });
+
+    it('can revoke the entire permission when a requiredScope is removed', () => {
+      const ethereumGoerliCaveat = {
+        requiredScopes: {
+          'eip155:1': {
+            methods: ['eth_call'],
+            notifications: ['chainChanged'],
+          },
+        },
+        optionalScopes: {
+          'eip155:5': {
+            methods: ['eth_call'],
+            notifications: ['accountsChanged'],
+          },
+        },
+        sessionProperties: {},
+        isMultichainOrigin: true,
+      };
+      const result = removeScope('eip155:1', ethereumGoerliCaveat);
+      expect(result).toStrictEqual({
+        operation: CaveatMutatorOperation.RevokePermission,
+      });
+    });
```

Comment:
I'm not sure we want this behavior yet? Since we're not enforcing required yet.

Probably doesn't need to be addressed in this migration PR, but just wanted to highlight it while I'm catching it.

Any thing I'm missing on this @jiexi @shanejonas ?

Reply:
you're right. There are two ways to address this:

1. Make a change in this mutator to no longer make this check against the requiredScopes
2. In the `wallet_createSession` handler, save everything in requiredScopes into optionalScopes

Reply:
could also just make changes in both spots

Reply:
> Make a change in this mutator to no longer make this check against the requiredScopes

I think this one. Since we may eventually enforce requiredScopes.

Reply:
https://github.com/MetaMask/core/pull/4784/commits/8415be6ec267fcd4f8c9f9df77bb2adaf6f0a476